File: Problems/0001_0099/0024_Swap_Nodes_in_Pairs/Project_CS/Swap_Nodes_in_Pairs.cs
using System;

//Definition for singly-linked list.
public class ListNode {
    public int val;
    public ListNode next;
    public ListNode(int x) { val = x; }
}

public class Solution
{
    public ListNode SwapPairs(ListNode head)
    {
        ListNode node = head;
        int temp;
        while (node != null)
        {
            if (node.next == null)
                break;
            temp = node.val;
            node.val = node.next.val;
            node.next.val = temp;
            if (node.next.next == null)
                break;
            node = node.next.next;
        }
        return head;
    }

    public int[] str_to_int_array(string s)
    {
        if (s.Length <= 0)
            return null;

        string[] flds = s.Split(',');
        int[] nums = new int[flds.Length];

        if (flds.Length <= 0)
            return nums;

        for (int i = 0; i < nums.Length; ++i)
        {
            nums[i] = int.Parse(flds[i]);
        }

        return nums;
    }

    public string output_int_array(int[] nums)
    {
        if (nums == null)
            return "";

        if (nums.Length <= 0)
            return "";

        string resultStr = "[" +  nums[0].ToString();
 
        for (int i = 1; i < nums.Length; ++i)
        {
            resultStr += "," + nums[i].ToString();
        }

        return resultStr + "]";
    }

    public void Main(string args)
    {
        string arg_str = args.Replace("[","").Replace("]","").Trim();
        int[] nums = str_to_int_array(arg_str);
        Console.WriteLine("nums = " + output_int_array(nums));

        Operate_ListNode ope = new Operate_ListNode();
        ListNode head = ope.set_ListNode(nums);
        Console.WriteLine("head = " + ope.output_ListNode(head));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        ListNode result = SwapPairs(head);

        sw.Stop();
        Console.WriteLine("result = " + ope.output_ListNode(result));
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}

File: Problems/0001_0099/0024_Swap_Nodes_in_Pairs/Project_Python3/Swap_Nodes_in_Pairs.py
import os
import sys
import time

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def swapPairs(self, head: 'ListNode') -> 'ListNode':
        node = head
        while node != None:
            if node.next == None:
                break
            temp = node.val
            node.val = node.next.val
            node.next.val = temp
            if node.next.next == None:
                break
            node = node.next.next

        return head

def set_nodes(nums, index):
    if nums == None:
        return None
    if index >= len(nums):
        return None
    
    node = ListNode(nums[index])
    node.next = set_nodes(nums, index + 1)

    return node

def output_nodes(ll):
    if ll == None:
        return ""

    retStr = str(ll.val)
    if ll.next != None:
        retStr += " -> " + output_nodes(ll.next)
    return retStr

def main():
    argv = sys.argv
    argc = len(argv)

    if (argc < 2):
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("args = %s" %temp)
        loop_main(temp)
    #    print("Hit Return to continue...")
    #    input()

def loop_main(temp):
    str_args = temp.replace("\"","").replace("[","").replace("]","").rstrip()

    nums = [int(val) for val in str_args.split(",")]
    print("nums = %s " %nums)

    head = set_nodes(nums, 0)
    print("head = %s " %output_nodes(head))

    time0 = time.time()

    sl = Solution()
    result = sl.swapPairs(head)

    time1 = time.time()

    print("result = %s" %output_nodes(result))
    print("Execute time ... : %f[s]\n" %(time1 - time0))

if __name__ == "__main__":
    main()
